### PR TITLE
openjdk (new formula)

### PR DESCRIPTION
**Pull request by VonUniGE**
Upgrade to jdk 10.0.2
The formula jdk.rb has been moved to jdk@9.rb
The alias to the current jdk has been renamed from  Aliases/jdk@9 to Aliases/jdk@10.rb

### CODE REVIEW

**Comment by VonUniGE** [comment]
To avoid any license question, I've switched jdk10 and jdk9 to OpenJDK. I've also removed the jdk8 bottle.

**Comment by sjackman** [comment]
Hi, @VonUniGE. @iMichka, the maintainer of Linuxbrew/core, is currently away on vacation for two weeks. I'll wait until he's back to discuss the merits of Oracle JDK vs OpenJDK.

The transition from Java 8 to Java 9 was a bit rough, so much so that I've considered reverting the default version of `jdk` back to version 8 (with `jdk@9` available for those that want it). I'd like to discuss with Michka before changing the default version of `jdk`.

If you'd like to submit a PR that adds `jdk@10`, I'd be happy to merge that. For consistency with the `jdk` formula, it ought to be the Oracle version of JDK. You can if you like submit a formula for `openjdk@10`.

**Comment by sjackman** [comment]
My understanding of the Oracle license is that we do have a right to redistribute Oracle JDK.
See `C. LICENSE TO DISTRIBUTE SOFTWARE.` http://www.oracle.com/technetwork/java/javase/terms/license/index.html

**Comment by sjackman** [comment]
What are the differences between Oracle JDK and OpenJDK? Do Debian and Ubuntu distribute OpenJDK? Does any Linux distribution distribute Oracle JDK?

**Comment by VonUniGE** [comment]
The main difference between OpenJDK and Oracle JDK is the license. OpenJDK is under [GPLv2](https://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html) and Oracle JDK is subject to the [Oracle Binary Code License Agreement](http://www.oracle.com/technetwork/java/javase/terms/license/index.html). Oracle JDK is based on OpenJDK with the addition of the Java Web Start, the Java Plugin and some closed source third party components such as graphics rasterizers.

Linux distributions generally build their own OpenJDK but this is not a trivial task. I don't know any linux distribution distributing directly the Oracle JDK.

The [official docker image of the OpenJDK](https://github.com/docker-library/openjdk) is based either on the [OpenJDK distributed with Alpine Linux](https://github.com/docker-library/openjdk/blob/master/8/jdk/alpine/Dockerfile) or on a [community build](https://github.com/ojdkbuild/ojdkbuild/) for the Windows version.

According to the [official Dockerfile of the Alpine OpenJDK](https://github.com/docker-library/openjdk/blob/master/8/jdk/alpine/Dockerfile):

```
# A few reasons for installing distribution-provided OpenJDK:
#
#  1. Oracle.  Licensing prevents us from redistributing the official JDK.
#
#  2. Compiling OpenJDK also requires the JDK to be installed, and it gets
#     really hairy.
#
#     For some sample build times, see Debian's buildd logs:
#       https://buildd.debian.org/status/logs.php?pkg=openjdk-8 
```
I'm not part of the Linuxbrew maintainers but personally, I think it is legally safer and ethically better to use only the OpenJDK and remove all remaining OracleJDK bottles.

**Comment by sjackman** [comment]
A few formulae in Brewsci/bio are broken because they don't work with Java 9+ and require Java 8. I haven't yet run into any tools that require Java 9+. I'd like to make the default version of `openjdk` be version 8. 

Would you like to submit formulae for `openjdk` (version 8) and `openjdk@10`?

These formulae can live in parallel with the existing Oracle `jdk` formula. Perhaps we can remove the Oracle formulae in the future.

**Comment by VonUniGE** [comment]
For now, it makes sense to have jdk@8 and openjdk@8 as default. Until now I didn't see, neither, any applications requiring jdk9 or jdk10 but this will change in the future.

I'm OK to write formulas for openjdk@8, openjdk@9, openjdk@10.

**Comment by sjackman** [comment]
Thanks, @VonUniGE! Please update this PR with three commits, one commit per new formula.

**Comment by VonUniGE** [comment]
Would you like also to change the default to jdk@8 and openjdk@8 on this PR?

**Comment by sjackman** [comment]
Not on this same PR. Let's keep this PR to adding `openjdk` and friends.

**Comment by VonUniGE** [comment]
Here is the three commits for openjdk@8, openjdk@9 and openjdk@10.

File: Formula/openjdk.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  if OS.mac?
+    url "http://java.com/"
+  else
+    url "https://github.com/ojdkbuild/contrib_jdk8u-ci/releases/download/jdk8u181-b13/jdk-8u181-ojdkbuild-linux-x64.zip"
+    sha256 "fe5f5f8870e3195b0ee4c25c597b990ebbe8e667f3a345ff0afc49a8ff212dae"
+  end
+
+  bottle :unneeded
+
+  keg_only :versioned_formula
```
Comment (sjackman) [inline Formula/openjdk.rb]: Please remove this line for this unversioned formula `openjdk`.

File: Formula/openjdk.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+  def install
+    odie "Use 'brew cask install java' on Mac OS" if OS.mac?
```
Comment (sjackman) [inline Formula/openjdk.rb]: Remove this line and add `depends_on :linux` above in the dependencies (outside of `def install`).

File: Formula/openjdk@10.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
+class OpenjdkAT10 < Formula
+	desc "Java Development Kit"
+	homepage "http://jdk.java.net/10/"
```
Comment (sjackman) [inline Formula/openjdk@10.rb]: Please replace these tab characters with two spaces.

File: Formula/openjdk@10.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  # tag "linuxbrew"
+
+  version "10.0.2"
+  version_scheme 1
```
Comment (sjackman) [inline Formula/openjdk@10.rb]: Please remove this line.

File: Formula/openjdk@10.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  if OS.mac?
+    url "http://java.com/"
+  else
+    url "https://download.java.net/java/GA/jdk10/10.0.2/19aef61b38124481863b1413dce1855f/13/openjdk-10.0.2_linux-x64_bin.tar.gz"
+    sha256 "f3b26abc9990a0b8929781310e14a339a7542adfd6596afb842fa0dd7e3848b2"
+  end
+
+  bottle do
+    cellar :any_skip_relocation
+  end
```
Comment (sjackman) [inline Formula/openjdk@10.rb]: Please remove these four lines.

File: Formula/openjdk@10.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+  def install
+    odie "Use 'brew cask install java' on Mac OS" if OS.mac?
```
Comment (sjackman) [inline Formula/openjdk@10.rb]: Please replace this line with `depends_on :linux`.

File: Formula/openjdk@10.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    prefix.install Dir["*"]
```
Comment (sjackman) [inline Formula/openjdk@10.rb]: Does this need the same `man` fixup?

File: Formula/openjdk@9.rb [hunk 8]
```diff
@@ -0,0 +1,38 @@
+class OpenjdkAT9 < Formula
+  desc "Java Development Kit"
+  homepage "http://jdk.java.net/archive/"
+  # tag "linuxbrew"
+
+  version "9.0.4"
+  version_scheme 1
```
Comment (sjackman) [inline Formula/openjdk@9.rb]: Please remove this line.

File: Formula/openjdk@10.rb [hunk 9]
```diff
@@ -0,0 +1,38 @@
+class OpenjdkAT10 < Formula
+	desc "Java Development Kit"
+	homepage "http://jdk.java.net/10/"
+  # tag "linuxbrew"
+
+  version "10.0.2"
```
Comment (sjackman) [inline Formula/openjdk@10.rb]: Is this line necessary? Is the version detected correctly without it? What is the detected version?

File: Formula/openjdk@9.rb [hunk 10]
```diff
@@ -0,0 +1,38 @@
+class OpenjdkAT9 < Formula
+  desc "Java Development Kit"
+  homepage "http://jdk.java.net/archive/"
+  # tag "linuxbrew"
+
+  version "9.0.4"
```
Comment (sjackman) [inline Formula/openjdk@9.rb]: Is this line necessary? What is the detected version without it?

File: Formula/openjdk@9.rb [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+  version_scheme 1
+  if OS.mac?
+    url "http://java.com/"
+  else
+	  url "https://download.java.net/java/GA/jdk9/9.0.4/binaries/openjdk-9.0.4_linux-x64_bin.tar.gz"
```
Comment (sjackman) [inline Formula/openjdk@9.rb]: Please replace this tab and spaces with four spaces. I recommend configuring your text editor to highlight tab characters.

File: Formula/openjdk@9.rb [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    sha256 "39362fb9bfb341fcc802e55e8ea59f4664ca58fd821ce956d48e1aa4fb3d2dec"
+  end
+
+  bottle do
+    cellar :any_skip_relocation
+  end
+
+  def install
+    odie "Use 'brew cask install java' on Mac OS" if OS.mac?
```
Comment (sjackman) [inline Formula/openjdk@9.rb]: Please replace this line with `depends_on :linux` outside `def install`.

File: Formula/openjdk@9.rb [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    prefix.install Dir["*"]
```
Comment (sjackman) [inline Formula/openjdk@9.rb]: Is the `man` fixup needed?

File: Formula/openjdk.rb (hunk 1 above)
Comment (VonUniGE) [reply Formula/openjdk.rb]: done

File: Formula/openjdk.rb (hunk 2 above)
Comment (VonUniGE) [reply Formula/openjdk.rb]: done

File: Formula/openjdk@10.rb (hunk 3 above)
Comment (VonUniGE) [reply Formula/openjdk@10.rb]: done

File: Formula/openjdk@10.rb (hunk 4 above)
Comment (VonUniGE) [reply Formula/openjdk@10.rb]: done

File: Formula/openjdk@10.rb (hunk 5 above)
Comment (VonUniGE) [reply Formula/openjdk@10.rb]: done

File: Formula/openjdk@9.rb (hunk 11 above)
Comment (VonUniGE) [reply Formula/openjdk@9.rb]: done. My editor is in a container, but yes I have to configure it better.

File: Formula/openjdk@9.rb (hunk 10 above)
Comment (VonUniGE) [reply Formula/openjdk@9.rb]: Not sure about this.

File: Formula/openjdk@9.rb (hunk 12 above)
Comment (VonUniGE) [reply Formula/openjdk@9.rb]: done

**Comment by VonUniGE** [comment]
I have to leave the keyboard for a few days. I'll commit when I get back.

**Comment by sjackman** [comment]
Enjoy your weekend!

**Comment by sjackman** [comment]
```
* Formula has other versions so create a versioned alias:
      cd /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Aliases
      ln -s ../Formula/openjdk.rb openjdk@1
  * Formula has invalid versioned aliases:
      openjdk@8
```
The alias symlink should probably be `openjdk@1.8` rather than `openjdk@8`.

File: Formula/openjdk@9.rb (hunk 10 above)
Comment (VonUniGE) [reply Formula/openjdk@9.rb]: When removing the line the `brew info` reports:
```
openjdk@9: stable 9.0.4_linux
````
With the explicit version, brew info reports:
```
openjdk@9: stable 9.0.4
```
I think that using the explicit version is more correct.

File: Formula/openjdk@10.rb (hunk 9 above)
Comment (VonUniGE) [reply Formula/openjdk@10.rb]: When removing the line the `brew info` reports:
```
openjdk@9: stable 10.0.2_linux
````
With the explicit version, brew info reports:
```
openjdk@9: stable 10.0.2
```
I think that using the explicit version is more correct.

File: Formula/openjdk@9.rb (hunk 8 above)
Comment (VonUniGE) [reply Formula/openjdk@9.rb]: done

File: Formula/openjdk@9.rb (hunk 13 above)
Comment (VonUniGE) [reply Formula/openjdk@9.rb]: No, manual pages are correctly installed without it.

**Comment by VonUniGE** [comment]
It's probably better to use `openjdk@8` as an alias instead of `openjdk@1.8`. It allows to stay coherent with the existing `jdk@7`, `jdk@8`formulas and the `jdk@9` alias. Moreover the version scheme 7, 8 ,9, 10 is the one commonly used by books, articles and users.

**Comment by VonUniGE** [comment]
I've integrated your suggestions and pushed one commit by new formula. I've also made some minor change to have three almost identical formulas.

File: Formula/openjdk@9.rb (hunk 10 above)
Comment (sjackman) [reply Formula/openjdk@9.rb]: I agree.

Comment (sjackman) [reply Formula/openjdk@9.rb]: I agree.

**Comment by sjackman** [comment]
Looks good to me. Thanks, @VonUniGE!

@iMichka Could you please review this PR when you get a chance?

File: Formula/openjdk.rb [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+    url "https://github.com/ojdkbuild/contrib_jdk8u-ci/releases/download/jdk8u181-b13/jdk-8u181-ojdkbuild-linux-x64.zip"
+    sha256 "fe5f5f8870e3195b0ee4c25c597b990ebbe8e667f3a345ff0afc49a8ff212dae"
+  end
+
+  depends_on :linux
+
+  def install
+    prefix.install Dir["*"]
+    share.mkdir
+    share.install prefix/"man"
```
Comment (sjackman) [inline Formula/openjdk.rb]: The following should work…
```ruby
share.install "man"
prefix.install Dir["*"]
```
and ditto for the other formulae.

**Comment by sjackman** [comment]
This PR looks good to me. Thanks, @VonUniGE !

A heads up that another contributor is working on building `openjdk` from source and contributing that formula upstream to Homebrew/core. When/if that happens, we can adopt the from-source build. In the mean time, I'm happy with binaries.

**Comment by iMichka** [comment]
Done with 28fc7114f72913baff794ea15b18bc03f8a63f97.

**Comment by iMichka** [comment]
I will build the bottles separately.

FYI I have some experience with building java/openjdk from source. You do not want to go down that road. I did that in my previous job, and you can spend days hacking around to get it done properly. I recommend having a look at the debian packages, you will understand :)

**Comment by iMichka** [comment]
Thanks @VonUniGE !

**Pull request closed without merging**